File: python/algorithms/mergesort.py
# Merge Sort
# Time Complexity: O(n log(n))
# Space Complexity: O(n)


def merge(arr, low, mid, high):
    n1 = mid - low + 1
    n2 = high - mid

    L = [0] * n1
    R = [0] * n2

    for i in range(n1):
        L[i] = arr[low + i]

    for j in range(n2):
        R[j] = arr[mid + 1 + j]

    i = 0
    j = 0
    k = low

    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1

    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1


def mergesort(arr, low, high):
    if low < high:
        mid = low + (high - low) // 2
        mergesort(arr, low, mid)
        mergesort(arr, mid + 1, high)
        merge(arr, low, mid, high)


if __name__ == '__main__':
    arr = [8, 1, 0, 2, 3, 4, 5, 7]
    mergesort(arr, 0, len(arr) - 1)
    assert arr == [0, 1, 2, 3, 4, 5, 7, 8]

File: c++/algorithms/sort/merge_sort.cpp
#include <iostream>

void merge(int A[], int lo, int mid, int hi)
{
    int n1 = mid - lo + 1;
    int n2 = hi - mid;
    int L[n1];
    int R[n2];
    int i, j, k;
    for (i = 0; i < n1; i++) {
        L[i] = A[lo + i];
    }
    for (j = 0; j < n2; j++) {
        R[j] = A[mid + 1 + j];
    }
    i = j = 0;
    k = lo;
    while (i < n1 && j < n2) {
        if (L[i] < R[j]) {
            A[k] = L[i];
            i++;
        } else {
            A[k] = R[j];
            j++;
        }
        k++;
    }
    while (i < n1) {
        A[k] = L[i];
        i++;
        k++;
    }
    while (j < n2) {
        A[k] = R[j];
        j++;
        k++;
    }
}

void merge_sort(int A[], int lo, int hi)
{
    if (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        merge_sort(A, lo, mid);
        merge_sort(A, mid + 1, hi);
        merge(A, lo, mid, hi);
    }
}

void print_arr(int A[], int n)
{
    for (int i = 0; i < n; i++) {
        std::cout << A[i] << " ";
    }
    std::cout << std::endl;
}

int main()
{
    int freq[] {2, 4, 9, 10, 0, 2, 7, 8, 18, 2002};
    int n = sizeof(freq) / sizeof(freq[0]);
    merge_sort(freq, 0, n - 1);
    print_arr(freq, n);
    return 0;
}

